src/full_client.toit

```diff
@@ -0,0 +1,689 @@
+  A mutex to queue the senders.
+
+  Relies on the fact that Toit mutexes are fair and execute them in the order
+    in which they reached the mutex.
+  */
+  sending_ /monitor.Mutex := monitor.Mutex
+
+  /**
+  The persistence store used by this client.
+
+  All messages with qos=1 are stored in this store immediately after they have
+    been sent. Once the client receives an 'ack' they are removed from it.
+
+  If the client closes, but the store isn't empty, then some messages might not
+    have reached the broker.
+
+  Use $PersistenceStore.do to iterate over the stored messages.
+  */
+  persistence_store /PersistenceStore
+
+  /**
+  Keeps track of the last acked packet.
+
+  If the packet was not acked during the callback of $handle we will do so automatically
+    when the callback returns.
+  */
+  unacked_packet_ /Packet? := null
+
+  /**
+  Constructs an MQTT client.
+
+  The client starts disconnected. Call $connect, followed by $handle to initiate
+    the connection.
+  */
+  constructor
+      --transport /Transport
+      --logger /log.Logger = log.default
+      --persistence_store /PersistenceStore? = null:
+    transport_ = ActivityMonitoringTransport.private_(transport)
+    logger_ = logger
+    this.persistence_store = persistence_store or MemoryPersistenceStore
+
+  /**
+  Checks whether a user is allowed to send a message.
+
+  The client must have connected, and run the $handle loop.
+  */
+  check_allowed_to_send_:
+    if is_closed: throw CLIENT_CLOSED_EXCEPTION
+    // The client is only active once $handle has been called.
+    if not is_running: throw "INVALID_STATE"
+
+  /**
+  Connects the client to the broker.
+
+  Once the client is connected, the user should call $handle to handle incoming
+    packets. Only, when the handler is active, can messages be sent.
+  */
+  connect -> none
+      --options /SessionOptions
+      --reconnection_strategy /ReconnectionStrategy? = null:
+    if state_ != STATE_CREATED_: throw "INVALID_STATE"
+    assert: not connection_
+
+    if reconnection_strategy:
+      reconnection_strategy_ = reconnection_strategy
+    else if options.clean_session:
+      reconnection_strategy_ = DefaultCleanSessionReconnectionStrategy
+    else:
+      reconnection_strategy_ = DefaultSessionReconnectionStrategy
+
+    session_ = Session_ options
+
+    state_ = STATE_CONNECTING_
+    reconnect_ --is_initial_connection
+    state_ = STATE_CONNECTED_
+
+  /**
+  Runs the receiving end of the client.
+
+  The client is not considered started until this method has been called.
+  This method is blocking.
+
+  The given $on_packet block is called for each received packet.
+    The block should $ack the packet as soon as possible.
+
+  The $when_running method executes a given block when this method is running.
+  */
+  handle [on_packet] -> none:
+    if state_ != STATE_CONNECTED_: throw "INVALID_STATE"
+    state_ = STATE_HANDLING_
+
+    try:
+      handling_latch_.set true
+      while true:
```

Comment:
Can the task that runs this get canceled? If so, we should probably leave the loop.

Reply:
Not sure of all the implications but added a check that `not task.is_canceled`.